notebooks/notebook_template_review.py

```diff
@@ -193,148 +198,141 @@ def parse_notebook(path: str) -> None:
-        OverviewRule().validate(path, cells)
+        cells = OverviewRule().validate(cells)
        
-        ObjectiveRule().validate(path, cells)
-        if NotebookRule.desc != '':
-            add_index(path, 
-                      tag, 
-                      NotebookRule.title, 
-                      NotebookRule.desc, 
-                      NotebookRule.uses, 
-                      NotebookRule.steps, 
-                      NotebookRule.git_link, 
-                      NotebookRule.colab_link, 
-                      NotebookRule.workbench_link
-            )
+        cells, desc, uses, steps, costs = ObjectiveRule().validate(cells)
             
-        RecommendationsRule().validate(path, cells)
+        cells = RecommendationsRule().validate(cells)
 
-        DatasetRule().validate(path, cells)
+        cells = DatasetRule().validate(cells)
 
-        CostsRule().validate(path, cells)
+        cells = CostsRule().validate(cells, costs)
 
-        SetupLocalRule().validate(path, cells)
+        cells = SetupLocalRule().validate(cells)
 
-        HelpersRule().validate(path, cells)
+        cells = HelpersRule().validate(cells)
 
-        InstallationRule().validate(path, cells)
+        cells = InstallationRule().validate(cells)
 
-        RestartRule().validate(path, cells)
-        
-        VersionsRule().validate(path, cells)
-        
-        BeforeBeginRule().validate(path, cells)
-        
-        EnableAPIsRule().validate(path, cells)
+        cells = RestartRule().validate(cells)
+
+        cells = VersionsRule().validate(cells)
+
+        cells = BeforeBeginRule().validate(cells)
+
+        cells = EnableAPIsRule().validate(cells)
         
-        SetupProjectRule().validate(path, cells)
+        cells = SetupProjectRule().validate(cells)
+  
+        if desc != '':
+            add_index(path, 
+                      tag, 
+                      linkback,
+                      title, 
+                      desc, 
+                      uses, 
+                      steps, 
+                      git_link, 
+                      colab_link, 
+                      workbench_link
+            )
 
 
 class NotebookRule(ABC):
     """
     Abstract class for defining notebook conformance rules
     """
-    cell_index = 0
-    title = ''
-    git_link = None
-    colab_link = None
-    workbench_link = None
-    desc = ''
-    uses = ''
-    steps = ''
-    costs = []
-    
-    @staticmethod
-    def init():
-        NotebookRule.cell_index = 0
-        NotebookRule.title = ''
-        NotebookRule.git_link = None
-        NotebookRule.colab_link = None
-        NotebookRule.workbench_link = None
-        NotebookRule.desc = ''
-        NotebookRule.uses = ''
-        NotebookRule.steps = ''
-        NotebookRule.costs = ''
-    
+    path = None     # The relative path to the notebook
+
     @abstractmethod
-    def validate(self, path: str, cells: list) -> None:
+    def validate(self, cells: list) -> list:
```

Comment:
Can you add a comment on what this does?

Should be:

```
def validate(self, cells: list) -> List[SomeTypeHere]:
```

Reply:
added